README.md
# bevy_mod_outline

[![crates.io](https://img.shields.io/crates/v/bevy_mod_outline.svg)](https://crates.io/crates/bevy_mod_outline)
[![docs](https://docs.rs/bevy_mod_outline/badge.svg)](https://docs.rs/bevy_mod_outline)

![Screenshot of bevy_mod_outline's shapes example](https://github.com/bevyengine/bevy-assets/blob/main/Assets/3D/bevy_mod_outline.png?raw=true)

This crate provides a Bevy plugin for drawing outlines around meshes using the
vertex extrusion method.

## Dependency

```toml
[dependencies]
bevy_mod_outline = "0.4"
```

## Examples

A rotating cube and torus with opaque and transparent outlines.

```shell
cargo run --example shapes
```

Multiple intersecting meshes sharing an outline plane. The outline stencil is offset to create
a gap between the object and its outline.

```shell
cargo run --example pieces
```

An outlined torus viewed through four cameras with different combinations of render layers
enabled.

```shell
cargo run --example render_layers
```

An animated jointed glTF model with an outline.

```shell
cargo run --example animated_fox
```

A glTF model with pre-baked outline normals.

```shell
cargo run --example hollow
```

## Versions

| This Version | Bevy version |
|--------------|--------------|
| 0.1.x        | 0.7.x        |
| 0.2.x        | 0.8.x        |
| 0.3.x        | 0.9.x        |
| 0.4.x        | 0.10.x       |
| 0.?.x        | 0.11.x       |

## Features

- `bevy_ui` _(default)_ - Adds a render graph edge to prevent clashing with the
UI. This adds a dependency on the `bevy_ui` crate and can be disabled if it is
not used.

## Licence

This crate is licensed under the Apache License, Version 2.0 (see
LICENCE-APACHE or <http://www.apache.org/licenses/LICENSE-2.0>) or the MIT
licence (see LICENCE-MIT or <http://opensource.org/licenses/MIT>), at your
option.

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Changelog

## bevy_mod_outline 0.?.? (2023-??-??)
- Updated Bevy dependency to 0.11.

## bevy_mod_outline 0.4.2 (2023-05-30)

### Fixed
- Fixed failures to propagate ComputedOutlineDepth when needed.
- Fixed Z-fighting between overlay and stencil with OpenGL wgpu back-end.

## bevy_mod_outline 0.4.1 (2023-04-11)

### Fixed
- Fixed panic if mesh has unused vertex indices.
- Fixed panic if the DepthMode hasn't propagated before rendering.

### Changed
- Changed normal weighting to use vertex rather than face normals if available.

## bevy_mod_outline 0.4.0 (2023-03-22)

### Added
- Added enabled flag to OutlineStencil.
- Added hollow example.

### Fixed
- Fixed outline depth propagating when inheritance not enabled.

### Changed
- Updated Bevy dependency to 0.10.
- Changed outline normal generator to use face normals.

## bevy_mod_outline 0.3.5 (2023-03-08)

### Fixed
- Fixed regression breaking SetOutlineDepth::Real.

## bevy_mod_outline 0.3.4 (2023-03-08)

### Fixed
- Fixed texture format error when HDR is enabled.
- Fixed bad clipping of outlines behind the camera.

## bevy_mod_outline 0.3.3 (2023-02-21)

### Fixed
- Fixed SetOutlineDepth being ignored in some cases (@mxgrey).
- Fixed defaulting to SetOutlineDepth::Real in some cases.
- Fixed missing examples from README.

## bevy_mod_outline 0.3.2 (2023-01-15)

### Added
- Added support for (Outline)RenderLayers components (@mxgrey).
- Added render_layers example.

## bevy_mod_outline 0.3.1 (2023-01-05)

### Added
- Added Lerp impls for OutlineVolume and OutlineStencil.
- Added Real mode to SetOutlineDepth enum.

## bevy_mod_outline 0.3.0 (2022-11-22)

### Added
- Added ComputedOutlineDepth, SetOutlineDepth, and InheritOutlineDepth.
- Added offset field to OutlineStencil.
- Added pieces example.

### Removed
- Removed align16 feature.

### Fixed
- Fixed errant debug println in AutoGenerateOutlineNormalsPlugin.

### Changed
- Updated Bevy dependency to 0.9.
- Renamed Outline component to OutlineVolume.

## bevy_mod_outline 0.2.5 (2023-01-14)

### Added
- Added support for (Outline)RenderLayers components (@mxgrey) (back-port from 0.3.2).
- Added render_layers example (back-port from 0.3.2).

### Fixed
- Fixed errant debug println in AutoGenerateOutlineNormalsPlugin (back-port from 0.3.0).

## bevy_mod_outline 0.2.4 (2022-10-12)

### Fixed
- Fixed adding outlines to jointed (skinned) meshes.

## bevy_mod_outline 0.2.3 (2022-08-28)

### Added
- Added AutoGenerateOutlineNormalsPlugin.

## bevy_mod_outline 0.2.2 (2022-08-23)

### Added
- Added a feature flag to control the dependency on bevy_ui (@Shatur).
- Added a feature flag to control uniform struct alignment.

### Fixed
- Fixed compilation on 32-bit platforms.
- Fixed insufficient alignment causing errors with WebGL.

### Changed
- Removed dependency on bevy's monolithic render feature flag (@Shatur).

## bevy_mod_outline 0.2.1 (2022-08-10)

### Added
- Added a constant to expose the outline pass node name.

### Fixed
- Fixed outlines causing UI to disappear when MSAA is enabled.
- Fixed bad derive allowing OutlineBundle to be inserted as a component.

## bevy_mod_outline 0.2.0 (2022-08-09)

### Added
- Added support for outline normals.

### Changed
- Updated Bevy dependency to 0.8.
- Changed from rendering in main pass to separate pass.
- Changed from using asset handles to plain components.

## bevy_mod_outline 0.1.0 (2022-06-14)

- Initial release


src/lib.rs
//! This crate provides a Bevy plugin, [`OutlinePlugin`], and associated components for
//! rendering outlines around meshes using the vertex extrusion method.
//!
//! Outlines are rendered in a seperate pass following the main 3D pass. The effect of this
//! pass is to present the outlines in depth sorted order according to the model translation
//! of each mesh. This ensures that outlines are not clipped by other geometry.
//!
//! The [`OutlineVolume`] component will, by itself, cover the original object entirely with
//! the outline colour. The [`OutlineStencil`] component must also be added to prevent the body
//! of an object from being filled it. This must be added to any entity which needs to appear on
//! top of an outline.
//!
//! The [`OutlineBundle`] and [`OutlineStencilBundle`] bundles can be used to add the right
//! components, including the required [`ComputedOutlineDepth`] component. Optionally, the
//! [`SetOutlineDepth`] and [`InheritOutlineDepth`] components may also be added to control the
//! depth ordering of outlines.
//!
//! Vertex extrusion works best with meshes that have smooth surfaces. To avoid visual
//! artefacts when outlining meshes with hard edges, see the
//! [`OutlineMeshExt::generate_outline_normals`] function and the
//! [`AutoGenerateOutlineNormalsPlugin`].

use bevy::asset::load_internal_asset;
use bevy::ecs::query::QueryItem;
use bevy::prelude::*;
use bevy::render::extract_component::{
    ExtractComponent, ExtractComponentPlugin, UniformComponentPlugin,
};
use bevy::render::mesh::MeshVertexAttribute;
use bevy::render::render_graph::RenderGraph;
use bevy::render::render_phase::{sort_phase_system, AddRenderCommand, DrawFunctions};
use bevy::render::render_resource::{SpecializedMeshPipelines, VertexFormat};
use bevy::render::view::RenderLayers;
use bevy::render::{Render, RenderApp, RenderSet};
use bevy::transform::TransformSystem;
use interpolation::Lerp;

use crate::draw::{
    queue_outline_stencil_mesh, queue_outline_volume_mesh, DrawOutline, DrawStencil,
};
use crate::node::{OpaqueOutline, OutlineNode, StencilOutline, TransparentOutline};
use crate::pipeline::{OutlinePipeline, FRAGMENT_SHADER_HANDLE, OUTLINE_SHADER_HANDLE};
use crate::uniforms::{
    extract_outline_stencil_uniforms, extract_outline_volume_uniforms,
    queue_outline_stencil_bind_group, queue_outline_volume_bind_group, OutlineFragmentUniform,
    OutlineStencilUniform, OutlineVolumeUniform,
};
use crate::view_uniforms::{
    extract_outline_view_uniforms, queue_outline_view_bind_group, OutlineViewUniform,
};

mod computed;
mod draw;
mod generate;
mod node;
mod pipeline;
mod uniforms;
mod view_uniforms;

pub use computed::*;
pub use generate::*;

// See https://alexanderameye.github.io/notes/rendering-outlines/

/// The direction to extrude the vertex when rendering the outline.
pub const ATTRIBUTE_OUTLINE_NORMAL: MeshVertexAttribute =
    MeshVertexAttribute::new("Outline_Normal", 1585570526, VertexFormat::Float32x3);

/// Name of the render graph node which draws the outlines.
///
/// This node runs after the main 3D passes and before the UI pass. The name can be used to
/// add additional constraints on node execution order with respect to other passes.
pub const OUTLINE_PASS_NODE_NAME: &str = "bevy_mod_outline_node";

/// A component for stenciling meshes during outline rendering.
#[derive(Clone, Component)]
pub struct OutlineStencil {
    /// Enable rendering of the stencil
    pub enabled: bool,
    /// Offset of the stencil in logical pixels
    pub offset: f32,
}

impl Default for OutlineStencil {
    fn default() -> Self {
        OutlineStencil {
            enabled: true,
            offset: 0.0,
        }
    }
}

impl ExtractComponent for OutlineStencil {
    type Query = &'static OutlineStencil;
    type Filter = ();
    type Out = Self;

    fn extract_component(item: QueryItem<Self::Query>) -> Option<Self> {
        Some(item.clone())
    }
}

fn lerp_bool(this: bool, other: bool, scalar: f32) -> bool {
    if scalar <= 0.0 {
        this
    } else if scalar >= 1.0 {
        other
    } else {
        this | other
    }
}

impl Lerp for OutlineStencil {
    type Scalar = f32;

    fn lerp(&self, other: &Self, scalar: &Self::Scalar) -> Self {
        OutlineStencil {
            enabled: lerp_bool(self.enabled, other.enabled, *scalar),
            offset: self.offset.lerp(&other.offset, scalar),
        }
    }
}

/// A component for rendering outlines around meshes.
#[derive(Clone, Component, Default)]
pub struct OutlineVolume {
    /// Enable rendering of the outline
    pub visible: bool,
    /// Width of the outline in logical pixels
    pub width: f32,
    /// Colour of the outline
    pub colour: Color,
}

impl Lerp for OutlineVolume {
    type Scalar = f32;

    fn lerp(&self, other: &Self, scalar: &Self::Scalar) -> Self {
        OutlineVolume {
            visible: lerp_bool(self.visible, other.visible, *scalar),
            width: self.width.lerp(&other.width, scalar),
            colour: {
                let [r, g, b, a] = self
                    .colour
                    .as_linear_rgba_f32()
                    .lerp(&other.colour.as_linear_rgba_f32(), scalar);
                Color::rgba_linear(r, g, b, a)
            },
        }
    }
}

/// A component for specifying what layer(s) the outline should be rendered for.
#[derive(Component, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Deref, DerefMut, Default)]
pub struct OutlineRenderLayers(pub RenderLayers);

impl ExtractComponent for OutlineRenderLayers {
    type Query = (
        Option<&'static OutlineRenderLayers>,
        Option<&'static RenderLayers>,
    );
    type Filter = Or<(With<OutlineVolume>, With<OutlineStencil>)>;
    type Out = Self;

    fn extract_component(
        (outline_mask, object_mask): (Option<&OutlineRenderLayers>, Option<&RenderLayers>),
    ) -> Option<Self> {
        Some(
            outline_mask
                .copied()
                .or_else(|| object_mask.copied().map(OutlineRenderLayers))
                .unwrap_or_default(),
        )
    }
}

/// A bundle for rendering stenciled outlines around meshes.
#[derive(Bundle, Clone, Default)]
pub struct OutlineBundle {
    pub outline: OutlineVolume,
    pub stencil: OutlineStencil,
    pub plane: ComputedOutlineDepth,
}

/// A bundle for stenciling meshes in the outlining pass.
#[derive(Bundle, Clone, Default)]
pub struct OutlineStencilBundle {
    pub stencil: OutlineStencil,
    pub plane: ComputedOutlineDepth,
}

/// Adds support for rendering outlines.
pub struct OutlinePlugin;

impl Plugin for OutlinePlugin {
    fn build(&self, app: &mut App) {
        load_internal_asset!(
            app,
            OUTLINE_SHADER_HANDLE,
            "outline.wgsl",
            Shader::from_wgsl
        );
        load_internal_asset!(
            app,
            FRAGMENT_SHADER_HANDLE,
            "fragment.wgsl",
            Shader::from_wgsl
        );

        app.add_plugins(ExtractComponentPlugin::<OutlineStencil>::extract_visible())
            .add_plugins(ExtractComponentPlugin::<OutlineRenderLayers>::default())
            .add_plugins(UniformComponentPlugin::<OutlineStencilUniform>::default())
            .add_plugins(UniformComponentPlugin::<OutlineVolumeUniform>::default())
            .add_plugins(UniformComponentPlugin::<OutlineFragmentUniform>::default())
            .add_plugins(UniformComponentPlugin::<OutlineViewUniform>::default())
            .add_systems(
                PostUpdate,
                compute_outline_depth.after(TransformSystem::TransformPropagate),
            )
            .sub_app_mut(RenderApp)
            .init_resource::<DrawFunctions<StencilOutline>>()
            .init_resource::<DrawFunctions<OpaqueOutline>>()
            .init_resource::<DrawFunctions<TransparentOutline>>()
            .init_resource::<OutlinePipeline>()
            .init_resource::<SpecializedMeshPipelines<OutlinePipeline>>()
            .add_render_command::<StencilOutline, DrawStencil>()
            .add_render_command::<OpaqueOutline, DrawOutline>()
            .add_render_command::<TransparentOutline, DrawOutline>()
            .add_systems(ExtractSchedule, extract_outline_view_uniforms)
            .add_systems(ExtractSchedule, extract_outline_stencil_uniforms)
            .add_systems(ExtractSchedule, extract_outline_volume_uniforms)
            .add_systems(
                Render,
                sort_phase_system::<StencilOutline>.in_set(RenderSet::PhaseSort),
            )
            .add_systems(
                Render,
                sort_phase_system::<OpaqueOutline>.in_set(RenderSet::PhaseSort),
            )
            .add_systems(
                Render,
                sort_phase_system::<TransparentOutline>.in_set(RenderSet::PhaseSort),
            )
            .add_systems(
                Render,
                queue_outline_view_bind_group.in_set(RenderSet::Queue),
            )
            .add_systems(
                Render,
                queue_outline_stencil_bind_group.in_set(RenderSet::Queue),
            )
            .add_systems(
                Render,
                queue_outline_volume_bind_group.in_set(RenderSet::Queue),
            )
            .add_systems(Render, queue_outline_stencil_mesh.in_set(RenderSet::Queue))
            .add_systems(Render, queue_outline_volume_mesh.in_set(RenderSet::Queue));

        let world = &mut app.sub_app_mut(RenderApp).world;
        let node = OutlineNode::new(world);

        let mut graph = world.resource_mut::<RenderGraph>();

        let draw_3d_graph = graph
            .get_sub_graph_mut(bevy::core_pipeline::core_3d::graph::NAME)
            .unwrap();
        draw_3d_graph.add_node(OUTLINE_PASS_NODE_NAME, node);
        draw_3d_graph.add_slot_edge(
            draw_3d_graph.input_node().id,
            bevy::core_pipeline::core_3d::graph::input::VIEW_ENTITY,
            OUTLINE_PASS_NODE_NAME,
            OutlineNode::IN_VIEW,
        );

        // Run after main 3D pass, but before UI psss
        draw_3d_graph.add_node_edge(
            bevy::core_pipeline::core_3d::graph::node::MAIN_OPAQUE_PASS,
            OUTLINE_PASS_NODE_NAME,
        );
        #[cfg(feature = "bevy_ui")]
        draw_3d_graph.add_node_edge(
            OUTLINE_PASS_NODE_NAME,
            bevy::ui::draw_ui_graph::node::UI_PASS,
        );
    }
}


src/pipeline.rs
use std::borrow::Cow;

use bevy::pbr::{MAX_CASCADES_PER_LIGHT, MAX_DIRECTIONAL_LIGHTS};
use bevy::prelude::*;
use bevy::reflect::TypeUuid;
use bevy::render::render_resource::{
    BindGroupLayout, BindGroupLayoutDescriptor, BindGroupLayoutEntry, BindingType, BlendState,
    BufferBindingType, BufferSize, ColorTargetState, ColorWrites, CompareFunction, DepthBiasState,
    DepthStencilState, Face, FragmentState, FrontFace, MultisampleState, PolygonMode,
    PrimitiveState, PrimitiveTopology, ShaderDefVal, ShaderSize, ShaderStages, StencilState,
    TextureFormat, VertexState,
};
use bevy::render::renderer::RenderDevice;
use bevy::render::texture::BevyDefault;
use bevy::render::view::ViewTarget;
use bevy::{
    pbr::MeshPipeline,
    render::{
        mesh::MeshVertexBufferLayout,
        render_resource::{
            RenderPipelineDescriptor, SpecializedMeshPipeline, SpecializedMeshPipelineError,
        },
    },
};
use bitfield::{bitfield_bitrange, bitfield_fields};

use crate::uniforms::{
    DepthMode, OutlineFragmentUniform, OutlineStencilUniform, OutlineVolumeUniform,
};
use crate::view_uniforms::OutlineViewUniform;
use crate::ATTRIBUTE_OUTLINE_NORMAL;

pub(crate) const OUTLINE_SHADER_HANDLE: HandleUntyped =
    HandleUntyped::weak_from_u64(Shader::TYPE_UUID, 2101625026478770097);

pub(crate) const FRAGMENT_SHADER_HANDLE: HandleUntyped =
    HandleUntyped::weak_from_u64(Shader::TYPE_UUID, 12033806834125368121);

#[derive(Clone, Copy, PartialEq, Eq)]
pub(crate) enum PassType {
    Stencil = 1,
    Opaque = 2,
    Transparent = 3,
}

#[derive(Copy, Clone, PartialEq, Eq, Hash)]
pub(crate) struct PipelineKey(u32);
bitfield_bitrange! {struct PipelineKey(u32)}

impl PipelineKey {
    bitfield_fields! {
        u32;
        msaa_samples_minus_one, set_msaa_samples_minus_one: 5, 0;
        primitive_topology_int, set_primitive_topology_int: 8, 6;
        pass_type_int, set_pass_type_int: 10, 9;
        depth_mode_int, set_depth_mode_int: 12, 11;
        pub offset_zero, set_offset_zero: 13;
        pub hdr_format, set_hdr_format: 14;
        pub opengl_workaround, set_opengl_workaround: 15;
    }

    pub(crate) fn new() -> Self {
        PipelineKey(0)
    }

    pub(crate) fn with_msaa(mut self, msaa: Msaa) -> Self {
        self.set_msaa_samples_minus_one(msaa as u32 - 1);
        self
    }

    pub(crate) fn msaa(&self) -> Msaa {
        match self.msaa_samples_minus_one() + 1 {
            x if x == Msaa::Off as u32 => Msaa::Off,
            x if x == Msaa::Sample2 as u32 => Msaa::Sample2,
            x if x == Msaa::Sample4 as u32 => Msaa::Sample4,
            x if x == Msaa::Sample8 as u32 => Msaa::Sample8,
            x => panic!("Invalid value for Msaa: {}", x),
        }
    }

    pub(crate) fn with_primitive_topology(mut self, primitive_topology: PrimitiveTopology) -> Self {
        self.set_primitive_topology_int(primitive_topology as u32);
        self
    }

    pub(crate) fn primitive_topology(&self) -> PrimitiveTopology {
        match self.primitive_topology_int() {
            x if x == PrimitiveTopology::PointList as u32 => PrimitiveTopology::PointList,
            x if x == PrimitiveTopology::LineList as u32 => PrimitiveTopology::LineList,
            x if x == PrimitiveTopology::LineStrip as u32 => PrimitiveTopology::LineStrip,
            x if x == PrimitiveTopology::TriangleList as u32 => PrimitiveTopology::TriangleList,
            x if x == PrimitiveTopology::TriangleStrip as u32 => PrimitiveTopology::TriangleStrip,
            x => panic!("Invalid value for PrimitiveTopology: {}", x),
        }
    }

    pub(crate) fn with_pass_type(mut self, pass_type: PassType) -> Self {
        self.set_pass_type_int(pass_type as u32);
        self
    }

    pub(crate) fn pass_type(&self) -> PassType {
        match self.pass_type_int() {
            x if x == PassType::Stencil as u32 => PassType::Stencil,
            x if x == PassType::Opaque as u32 => PassType::Opaque,
            x if x == PassType::Transparent as u32 => PassType::Transparent,
            x => panic!("Invalid value for PassType: {}", x),
        }
    }

    pub(crate) fn with_depth_mode(mut self, depth_mode: DepthMode) -> Self {
        self.set_depth_mode_int(depth_mode as u32);
        self
    }

    pub(crate) fn depth_mode(&self) -> DepthMode {
        match self.depth_mode_int() {
            x if x == DepthMode::Flat as u32 => DepthMode::Flat,
            x if x == DepthMode::Real as u32 => DepthMode::Real,
            x => panic!("Invalid value for DepthMode: {}", x),
        }
    }

    pub(crate) fn with_offset_zero(mut self, offset_zero: bool) -> Self {
        self.set_offset_zero(offset_zero);
        self
    }

    pub(crate) fn with_hdr_format(mut self, hdr_format: bool) -> Self {
        self.set_hdr_format(hdr_format);
        self
    }

    pub(crate) fn with_opengl_workaround(mut self, opengl_workaround: bool) -> Self {
        self.set_opengl_workaround(opengl_workaround);
        self
    }
}

#[derive(Resource)]
pub(crate) struct OutlinePipeline {
    mesh_pipeline: MeshPipeline,
    pub outline_view_bind_group_layout: BindGroupLayout,
    pub outline_stencil_bind_group_layout: BindGroupLayout,
    pub outline_volume_bind_group_layout: BindGroupLayout,
}

impl FromWorld for OutlinePipeline {
    fn from_world(world: &mut World) -> Self {
        let world = world.cell();
        let mesh_pipeline = world.get_resource::<MeshPipeline>().unwrap().clone();
        let render_device = world.get_resource::<RenderDevice>().unwrap();
        let outline_view_bind_group_layout =
            render_device.create_bind_group_layout(&BindGroupLayoutDescriptor {
                label: Some("outline_view_bind_group_layout"),
                entries: &[BindGroupLayoutEntry {
                    binding: 0,
                    visibility: ShaderStages::VERTEX,
                    ty: BindingType::Buffer {
                        ty: BufferBindingType::Uniform,
                        has_dynamic_offset: true,
                        min_binding_size: BufferSize::new(OutlineViewUniform::SHADER_SIZE.get()),
                    },
                    count: None,
                }],
            });
        let outline_volume_bind_group_layout =
            render_device.create_bind_group_layout(&BindGroupLayoutDescriptor {
                label: Some("outline_volume_bind_group_layout"),
                entries: &[
                    BindGroupLayoutEntry {
                        binding: 0,
                        visibility: ShaderStages::VERTEX,
                        ty: BindingType::Buffer {
                            ty: BufferBindingType::Uniform,
                            has_dynamic_offset: true,
                            min_binding_size: BufferSize::new(
                                OutlineVolumeUniform::SHADER_SIZE.get(),
                            ),
                        },
                        count: None,
                    },
                    BindGroupLayoutEntry {
                        binding: 1,
                        visibility: ShaderStages::FRAGMENT,
                        ty: BindingType::Buffer {
                            ty: BufferBindingType::Uniform,
                            has_dynamic_offset: true,
                            min_binding_size: BufferSize::new(
                                OutlineFragmentUniform::SHADER_SIZE.get(),
                            ),
                        },
                        count: None,
                    },
                ],
            });
        let outline_stencil_bind_group_layout =
            render_device.create_bind_group_layout(&BindGroupLayoutDescriptor {
                label: Some("outline_stencil_bind_group_layout"),
                entries: &[BindGroupLayoutEntry {
                    binding: 0,
                    visibility: ShaderStages::VERTEX,
                    ty: BindingType::Buffer {
                        ty: BufferBindingType::Uniform,
                        has_dynamic_offset: true,
                        min_binding_size: BufferSize::new(OutlineStencilUniform::SHADER_SIZE.get()),
                    },
                    count: None,
                }],
            });
        OutlinePipeline {
            mesh_pipeline,
            outline_view_bind_group_layout,
            outline_stencil_bind_group_layout,
            outline_volume_bind_group_layout,
        }
    }
}

impl SpecializedMeshPipeline for OutlinePipeline {
    type Key = PipelineKey;

    fn specialize(
        &self,
        key: Self::Key,
        mesh_layout: &MeshVertexBufferLayout,
    ) -> Result<RenderPipelineDescriptor, SpecializedMeshPipelineError> {
        let mut targets = vec![];
        let mut bind_layouts = vec![if key.msaa() == Msaa::Off {
            self.mesh_pipeline.view_layout.clone()
        } else {
            self.mesh_pipeline.view_layout_multisampled.clone()
        }];
        let mut buffer_attrs = vec![Mesh::ATTRIBUTE_POSITION.at_shader_location(0)];
        let mut vertex_defs = vec![
            ShaderDefVal::Int(
                "MAX_CASCADES_PER_LIGHT".to_string(),
                MAX_CASCADES_PER_LIGHT as i32,
            ),
            ShaderDefVal::Int(
                "MAX_DIRECTIONAL_LIGHTS".to_string(),
                MAX_DIRECTIONAL_LIGHTS as i32,
            ),
        ];
        let mut fragment_defs = vec![];
        bind_layouts.push(
            if mesh_layout.contains(Mesh::ATTRIBUTE_JOINT_INDEX)
                && mesh_layout.contains(Mesh::ATTRIBUTE_JOINT_WEIGHT)
            {
                vertex_defs.push(ShaderDefVal::from("SKINNED"));
                buffer_attrs.push(Mesh::ATTRIBUTE_JOINT_INDEX.at_shader_location(2));
                buffer_attrs.push(Mesh::ATTRIBUTE_JOINT_WEIGHT.at_shader_location(3));
                self.mesh_pipeline.mesh_layouts.skinned.clone()
            } else {
                self.mesh_pipeline.mesh_layouts.model_only.clone()
            },
        );
        bind_layouts.push(self.outline_view_bind_group_layout.clone());
        let cull_mode;
        if key.depth_mode() == DepthMode::Flat {
            vertex_defs.push(ShaderDefVal::from("FLAT_DEPTH"));
            cull_mode = Some(Face::Back);
        } else if key.pass_type() == PassType::Stencil {
            cull_mode = Some(Face::Back);
        } else {
            cull_mode = Some(Face::Front);
        }
        if key.offset_zero() {
            vertex_defs.push(ShaderDefVal::from("OFFSET_ZERO"));
        } else {
            buffer_attrs.push(
                if mesh_layout.contains(ATTRIBUTE_OUTLINE_NORMAL) {
                    ATTRIBUTE_OUTLINE_NORMAL
                } else {
                    Mesh::ATTRIBUTE_NORMAL
                }
                .at_shader_location(1),
            );
        }
        match key.pass_type() {
            PassType::Stencil => {
                bind_layouts.push(self.outline_stencil_bind_group_layout.clone());
            }
            PassType::Opaque | PassType::Transparent => {
                fragment_defs.push(ShaderDefVal::from("VOLUME"));
                targets.push(Some(ColorTargetState {
                    format: if key.hdr_format() {
                        ViewTarget::TEXTURE_FORMAT_HDR
                    } else {
                        TextureFormat::bevy_default()
                    },
                    blend: Some(if key.pass_type() == PassType::Transparent {
                        BlendState::ALPHA_BLENDING
                    } else {
                        BlendState::REPLACE
                    }),
                    write_mask: ColorWrites::ALL,
                }));

                bind_layouts.push(self.outline_volume_bind_group_layout.clone());
            }
        }
        if key.opengl_workaround() {
            let val = ShaderDefVal::from("OPENGL_WORKAROUND");
            vertex_defs.push(val.clone());
            fragment_defs.push(val);
        }
        let buffers = vec![mesh_layout.get_layout(&buffer_attrs)?];
        Ok(RenderPipelineDescriptor {
            vertex: VertexState {
                shader: OUTLINE_SHADER_HANDLE.typed::<Shader>(),
                entry_point: "vertex".into(),
                shader_defs: vertex_defs,
                buffers,
            },
            fragment: Some(FragmentState {
                shader: FRAGMENT_SHADER_HANDLE.typed::<Shader>(),
                shader_defs: fragment_defs,
                entry_point: "fragment".into(),
                targets,
            }),
            layout: bind_layouts,
            primitive: PrimitiveState {
                front_face: FrontFace::Ccw,
                cull_mode,
                unclipped_depth: false,
                polygon_mode: PolygonMode::Fill,
                conservative: false,
                topology: key.primitive_topology(),
                strip_index_format: None,
            },
            depth_stencil: Some(DepthStencilState {
                format: TextureFormat::Depth32Float,
                depth_write_enabled: true,
                depth_compare: CompareFunction::Greater,
                stencil: StencilState::default(),
                bias: DepthBiasState {
                    constant: if key.depth_mode() == DepthMode::Flat
                        && key.pass_type() == PassType::Stencil
                    {
                        2 // 1 is empirically not enough to prevent Z-fighting.
                    } else {
                        0
                    },
                    ..default()
                },
            }),
            multisample: MultisampleState {
                count: key.msaa().samples(),
                mask: !0,
                alpha_to_coverage_enabled: false,
            },
            push_constant_ranges: default(),
            label: Some(Cow::Borrowed("outline_pipeline")),
        })
    }
}


examples/pieces.rs
use std::f32::consts::TAU;

use bevy::{
    prelude::{
        shape::{Capsule, Plane, Torus, UVSphere},
        *,
    },
    window::close_on_esc,
};

use bevy_mod_outline::*;

#[bevy_main]
fn main() {
    App::new()
        .insert_resource(Msaa::Sample4)
        .insert_resource(ClearColor(Color::BLACK))
        .add_plugins(DefaultPlugins)
        .add_plugins(OutlinePlugin)
        .add_systems(Startup, setup)
        .add_systems(Update, close_on_esc)
        .add_systems(Update, rotates)
        .run();
}

#[derive(Component)]
struct Rotates;

fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    let outline = OutlineBundle {
        outline: OutlineVolume {
            visible: true,
            colour: Color::WHITE,
            width: 10.0,
        },
        stencil: OutlineStencil {
            offset: 5.0,
            ..default()
        },
        ..default()
    };

    // Add sphere with child meshes sticking out of it
    commands
        .spawn(PbrBundle {
            mesh: meshes.add(
                UVSphere {
                    radius: 0.75,
                    sectors: 30,
                    stacks: 30,
                }
                .into(),
            ),
            material: materials.add(Color::rgb(0.9, 0.1, 0.1).into()),
            transform: Transform::from_translation(Vec3::new(0.0, 1.0, 0.0)),
            ..default()
        })
        .insert(outline.clone())
        .insert(Rotates)
        .with_children(|parent| {
            parent
                .spawn(PbrBundle {
                    mesh: meshes.add(
                        Capsule {
                            radius: 0.2,
                            rings: 15,
                            depth: 1.0,
                            latitudes: 15,
                            longitudes: 15,
                            ..Default::default()
                        }
                        .into(),
                    ),
                    material: materials.add(Color::rgb(0.1, 0.1, 0.9).into()),
                    transform: Transform::from_rotation(Quat::from_axis_angle(Vec3::X, TAU / 4.0))
                        .with_translation(Vec3::new(0.0, 0.0, 0.75)),
                    ..default()
                })
                .insert(outline.clone())
                .insert(InheritOutlineDepth);
            parent
                .spawn(PbrBundle {
                    mesh: meshes.add(
                        Torus {
                            radius: 0.5,
                            ring_radius: 0.1,
                            subdivisions_segments: 30,
                            subdivisions_sides: 15,
                        }
                        .into(),
                    ),
                    material: materials.add(Color::rgb(0.1, 0.1, 0.9).into()),
                    transform: Transform::from_rotation(Quat::from_axis_angle(Vec3::Z, TAU / 4.0))
                        .with_translation(Vec3::new(0.0, 0.0, -0.75)),
                    ..default()
                })
                .insert(outline.clone())
                .insert(InheritOutlineDepth);
        });

    // Add plane, light source, and camera
    commands.spawn(PbrBundle {
        mesh: meshes.add(Mesh::from(Plane {
            size: 5.0,
            subdivisions: 0,
        })),
        material: materials.add(Color::rgb(0.3, 0.5, 0.3).into()),
        ..default()
    });
    commands.spawn(PointLightBundle {
        point_light: PointLight {
            intensity: 1500.0,
            shadows_enabled: true,
            ..default()
        },
        transform: Transform::from_xyz(4.0, 8.0, 4.0),
        ..default()
    });
    commands.spawn(Camera3dBundle {
        transform: Transform::from_xyz(-2.0, 2.5, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
        ..default()
    });
}

fn rotates(mut query: Query<&mut Transform, With<Rotates>>, timer: Res<Time>) {
    for mut transform in query.iter_mut() {
        transform.rotate_axis(Vec3::Y, 0.75 * timer.delta_seconds());
    }
}
